api/nodes/permissions.py

```diff
@@ -63,6 +65,36 @@ def has_object_permission(self, request, view, obj):
             has_auth = parent_node.can_edit(auth)
             return has_auth
 
+
+class ContributorOrPublicForComments(permissions.BasePermission):
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, Comment, CommentReport)), 'obj must be a Node, Comment or CommentReport, got {}'.format(obj)
+        auth = get_user_auth(request)
+        if isinstance(obj, CommentReport):
+            comment_id = request.parser_context['kwargs']['comment_id']
+            comment = Comment.load(comment_id)
+            if not comment:
+                raise NotFound
+            node = comment.node
+        elif isinstance(obj, Comment):
+            comment = obj
+            node = obj.node
+        elif isinstance(obj, Node):
+            node = obj
+
+        view_name = view.get_view_name()
+        is_report_view = (view_name == 'Comment Reports List' or view_name == 'Comment Report Detail')
```

Comment:
This seems a big dangerous to me. The view name is programmatically generated based on the class name. If someone refactors the class and changes the name this will break. Could you use an isinstance check instead? That way if they do an automated refactor or even a search and replace it will catch these usages too.